### PR TITLE
fix(artifacts): update artifacts cache file permissions for NamedTemporaryFile

### PR DESCRIPTION
**Pull request by ibindlish**
Fixes
-----
- Fixes [WB-15992](https://wandb.atlassian.net/browse/WB-15992)
- Fixes #NNNN

Description
-----------
[What does the PR do?](https://weightsandbiases.slack.com/archives/C01KQ5KTDC3/p1696877521780069) - Customer Elucid facing issues.

Python's `NamedTemporaryFile()` creates files with `0o600` permissions, ignoring any system defaults that the users may have set. This PR updates the permissions of the cached files so that the permissions on downloaded artifact files are set according to expectations. 

Potentially caused by this PR: https://github.com/wandb/wandb/pull/6172

### <samp>🤖 Generated by Copilot at 789a02e</samp>

Fixed a bug with temporary file permissions in the artifacts cache and improved logging in the artifact manifest entry class. Changed files: `artifacts_cache.py` and `artifact_manifest_entry.py`.

Testing
-------
Artifact with directories
```
with wandb.init(entity=ENTITY, project="wandb") as run:
    artifact = run.use_artifact("estellazx/small tests/dir_artifact:v1", type="model")
    path = artifact.download()
```

Artifact with one file
```
run = wandb.init()
artifact = run.use_artifact('estellazx/small tests/add_one_file:v2', type='dataset')
artifact_dir = artifact.download()
```

Checklist
-------
- [ ] Include reference to internal ticket "Fixes WB-NNNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

### <samp>🤖 Generated by Copilot at 789a02e</samp>

> _To log messages with some flair_
> _We imported the `termlog` from `wandb.errors.term`_
> _But we also had to fix_
> _The permissions of temp files with a trick_
> _By using the `umask` to make them conform_

[WB-15992]: https://wandb.atlassian.net/browse/WB-15992?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by szymon-piechowicz-wandb — COMMENTED** [review]
For reference:
- what happens when we download a file: https://weightsandbiases.slack.com/archives/C04MNBGJDBN/p1696970230685539?thread_ts=1696878870.456309&cid=C04MNBGJDBN
- `NamedTemporaryFile` creates files with 600 mode: https://stackoverflow.com/questions/10541760/can-i-set-the-umask-for-tempfile-namedtemporaryfile-in-python

File: wandb/sdk/artifacts/artifacts_cache.py [hunk 1]
```diff
@@ -181,6 +181,13 @@ def helper(mode: str = "w") -> Generator[IO, None, None]:
             try:
                 yield temp_file
                 temp_file.close()
+                # The NamedTemporaryFile is always created with mode 0600, hardcoded at
+                # tempfile.py, line 235, because it is private to your process until
+                # you open it up with chmod. Here we change the permissions back to
+                # the system defaults, with a temporary umask assignment to 0o022
+                umask = os.umask(0o22)
+                os.umask(umask)
```
Comment (szymon-piechowicz-wandb) [inline wandb/sdk/artifacts/artifacts_cache.py:189]: Nit:
- let's just point to the StackOverflow discussion
- it wasn't clear from this comment why you set the umask twice
```suggestion
                # NamedTemporaryFile sets the file mode to 600 [1], we reset to the default.
                # [1] https://stackoverflow.com/questions/10541760/can-i-set-the-umask-for-tempfile-namedtemporaryfile-in-python
                umask = os.umask(0o22)  # we have to set umask to get the current value
                os.umask(umask)  # reset
```

File: wandb/sdk/artifacts/artifacts_cache.py [hunk 2]
```diff
@@ -181,6 +181,13 @@ def helper(mode: str = "w") -> Generator[IO, None, None]:
             try:
                 yield temp_file
                 temp_file.close()
+                # The NamedTemporaryFile is always created with mode 0600, hardcoded at
+                # tempfile.py, line 235, because it is private to your process until
+                # you open it up with chmod. Here we change the permissions back to
+                # the system defaults, with a temporary umask assignment to 0o022
+                umask = os.umask(0o22)
```
Comment (szymon-piechowicz-wandb) [inline wandb/sdk/artifacts/artifacts_cache.py:188]: This is not thread safe, see: https://stackoverflow.com/a/44130549

**Review by szymon-piechowicz-wandb — CHANGES_REQUESTED** [review]
This sets the cache file mode to the default. But we still add u+w g+w o+w after copying the file to the user dir [here](https://github.com/wandb/wandb/blob/7e1f49b5a69c1c5c91f96953617a6954f90f78de/wandb/sdk/lib/filesystem.py#L177).

File: wandb/sdk/artifacts/artifacts_cache.py (hunk 2 above)
Comment (szymon-piechowicz-wandb) [reply wandb/sdk/artifacts/artifacts_cache.py:188]: I think this is the only safe way to get the umask: https://stackoverflow.com/a/53357635

Comment (ibindlish) [reply wandb/sdk/artifacts/artifacts_cache.py]: updated

File: wandb/sdk/artifacts/artifacts_cache.py [hunk 3]
```diff
@@ -181,6 +182,11 @@ def helper(mode: str = "w") -> Generator[IO, None, None]:
             try:
                 yield temp_file
                 temp_file.close()
+                # NamedTemporaryFile sets the file mode to 600 [1], we reset to the default.
+                # [1] https://stackoverflow.com/questions/10541760/can-i-set-the-umask-for-tempfile-namedtemporaryfile-in-python
+                umask_cmd = ("python", "-c", "import os; print(os.umask(22))")
```
Comment (szymon-piechowicz-wandb) [inline wandb/sdk/artifacts/artifacts_cache.py:187]: `python` might not be on `PATH`, use:
```suggestion
                umask_cmd = (sys.executable, "-c", "import os; print(os.umask(22))")
```

**Review by szymon-piechowicz-wandb — APPROVED** [review]

**Pull request merged**